crypto/blake2/blake2b.c

```diff
@@ -0,0 +1,310 @@
+ *
+ * Copyright 2012, Samuel Neves <sneves@dei.uc.pt>.
+ * You may use this under the terms of the CC0, the OpenSSL Licence, or the
+ * Apache Public License 2.0, at your option.  The terms of these licenses can
+ * be found at:
+ *
+ * - OpenSSL license   : https://www.openssl.org/source/license.html
+ * - Apache 2.0        : http://www.apache.org/licenses/LICENSE-2.0
+ * - CC0 1.0 Universal : http://creativecommons.org/publicdomain/zero/1.0
+ *
+ * More information about the BLAKE2 hash function can be found at
+ * https://blake2.net.
+ *
+ */
+
+/* crypto/blake2/blake2b.c */
+
+#include <stdint.h>
+#include <string.h>
+#include <stdio.h>
+#include <openssl/crypto.h>
+
+#include "internal/blake2_locl.h"
+#include "blake2_impl.h"
+
+static const uint64_t blake2b_IV[8] =
+{
+    0x6a09e667f3bcc908ULL, 0xbb67ae8584caa73bULL,
+    0x3c6ef372fe94f82bULL, 0xa54ff53a5f1d36f1ULL,
+    0x510e527fade682d1ULL, 0x9b05688c2b3e6c1fULL,
+    0x1f83d9abfb41bd6bULL, 0x5be0cd19137e2179ULL
+};
+
+static const uint8_t blake2b_sigma[12][16] =
+{
+    {  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15 } ,
+    { 14, 10,  4,  8,  9, 15, 13,  6,  1, 12,  0,  2, 11,  7,  5,  3 } ,
+    { 11,  8, 12,  0,  5,  2, 15, 13, 10, 14,  3,  6,  7,  1,  9,  4 } ,
+    {  7,  9,  3,  1, 13, 12, 11, 14,  2,  6,  5, 10,  4,  0, 15,  8 } ,
+    {  9,  0,  5,  7,  2,  4, 10, 15, 14,  1, 11, 12,  6,  8,  3, 13 } ,
+    {  2, 12,  6, 10,  0, 11,  8,  3,  4, 13,  7,  5, 15, 14,  1,  9 } ,
+    { 12,  5,  1, 15, 14, 13,  4, 10,  0,  7,  6,  3,  9,  2,  8, 11 } ,
+    { 13, 11,  7, 14, 12,  1,  3,  9,  5,  0, 15,  4,  8,  6,  2, 10 } ,
+    {  6, 15, 14,  9, 11,  3,  0,  8, 12,  2, 13,  7,  1,  4, 10,  5 } ,
+    { 10,  2,  8,  4,  7,  6,  1,  5, 15, 11,  9, 14,  3, 12, 13 , 0 } ,
+    {  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15 } ,
+    { 14, 10,  4,  8,  9, 15, 13,  6,  1, 12,  0,  2, 11,  7,  5,  3 }
+};
+
+
+static inline void blake2b_set_lastnode(BLAKE2B_CTX *S)
+{
+    S->f[1] = -1;
+}
+
+/* Some helper functions, not necessarily useful */
+static inline void blake2b_set_lastblock(BLAKE2B_CTX *S)
+{
+    if(S->last_node) {
+        blake2b_set_lastnode(S);
+    }
+    S->f[0] = -1;
+}
+
+static inline void blake2b_increment_counter(BLAKE2B_CTX *S,
+                                             const uint64_t inc)
+{
+    S->t[0] += inc;
+    S->t[1] += (S->t[0] < inc);
+}
+
+static inline void blake2b_init0(BLAKE2B_CTX *S)
+{
+    int i;
+    memset(S, 0, sizeof(BLAKE2B_CTX));
+
+    for(i = 0; i < 8; ++i) {
+        S->h[i] = blake2b_IV[i];
+    }
+}
+
+/* init xors IV with input parameter block */
+static int blake2b_init_param(BLAKE2B_CTX *S, const BLAKE2B_PARAM *P)
+{
+    size_t i;
+    const uint8_t *p = (const uint8_t *)(P);
+    blake2b_init0(S);
+
+    /* The param struct is carefully hand packed, and should be 64 bytes on
+     * every platform. */
+    OPENSSL_assert(sizeof(BLAKE2B_PARAM) == 64);
```

Comment:
On Sun, Mar 6, 2016 at 8:37 AM, Kurt Roeckx <notifications@github.com>
wrote:

> In crypto/blake2/blake2b.c
> <https://github.com/openssl/openssl/pull/566#discussion_r55143610>:
>
> > +    /* The param struct is carefully hand packed, and should be 64 bytes on
> > +     * every platform. */
> > +    OPENSSL_assert(sizeof(BLAKE2B_PARAM) == 64);
>
> Can you turn this into a compile time rather than a runtime error?
>
I am open to suggestions, but I could not find a way to do so that is not
ugly.  For example, I could declare an array:

    size_t dummy[1-(sizeof(BLAKE2B_PARAM) - 64)*(sizeof(BLAKE2B_PARAM) -
64)]

And then actually use it so this does not generate an unused variable
warning.  This gives a compiler error due to 0 or negative array size at
compile time.  This could probably be done in a more readable way with some
sort of pragma, but only weird compilers are going to pack a structure of
64 byte variables in in a non-64 byte memory layout, and I would not expect
pragmas to work in those compilers.

I checked the generated assembly code.  The assertion is optimized away, so
this is not a speed issue.